### PR TITLE
[Availability Zone] VM Placement in a Zone integration

### PR DESCRIPTION
**Pull request by bgriddaluru**
Design Doc: https://github.com/microsoft/wssdagent/pull/994 (VM_Placement_in_a_Zone.md)

### CODE REVIEW

**Review by msftpgayam — APPROVED** [review]

File: services/compute/virtualmachine/virtualmachine.go [hunk 1]
```diff
@@ -506,6 +512,21 @@ func (c *client) getWssdAvailabilitySetReference(s *compute.AvailabilitySetRefer
 	return availabilitySet, nil
 }
 
+func (c *client) getWssdZoneConfiguration(zoneProfile *compute.ZoneConfiguration) (*wssdcommon.ZoneConfiguration, error) {
+	if zoneProfile == nil {
+		return nil, nil
+	}
+
+	zones := []*wssdcommon.ZoneReference{}
+	for _, zone := range *zoneProfile.Zones {
```
Comment (msftpgayam) [inline services/compute/virtualmachine/virtualmachine.go:521]: We should add a nil check here?

File: services/compute/virtualmachine/virtualmachine.go [hunk 2]
```diff
@@ -882,3 +904,21 @@ func (c *client) getVirtualMachineProxyConfiguration(proxyConfiguration *wssdclo
...
 }
+
+func (c *client) getZoneConfiguration(zoneConfiguration *wssdcommon.ZoneConfiguration) *compute.ZoneConfiguration {
+	if zoneConfiguration == nil || len(zoneConfiguration.Zones) == 0 {
+		return nil
+	}
+
+	zones := []compute.Zone{}
+	strinctPlacement := zoneConfiguration.StrictPlacement
+	for _, zone := range zoneConfiguration.Zones {
```
Comment (msftpgayam) [inline services/compute/virtualmachine/virtualmachine.go:915]: Should we add a nil check?

**Review by msftpgayam — CHANGES_REQUESTED** [review]

File: services/compute/virtualmachine/virtualmachine.go (hunk 1 above)
Comment (bgriddaluru) [reply services/compute/virtualmachine/virtualmachine.go:521]: Fixed in new patch

File: services/compute/virtualmachine/virtualmachine.go (hunk 2 above)
Comment (bgriddaluru) [reply services/compute/virtualmachine/virtualmachine.go:915]: zoneConfiguration.Zones is an array of pointers i.e. 'Zones   []*ZoneReference'. So nil check is not required.

**Review by msftpgayam — DISMISSED** [review]

**Review by piyushka17 — APPROVED** [review]

**Review by piyushka17 — DISMISSED** [review]

**Comment by msftpgayam** [comment]
This PR still has conflicts

**Review by msftpgayam — APPROVED** [review]

**Review by piyushka17 — APPROVED** [review]

**Pull request merged**